### PR TITLE
Feature: Add basic backend configurations

### PR DESCRIPTION
**Pull request by yhaliaw**
### Overview

Add the following charm configurations:

- health-check-interval
- backends-path
- proxy-cache-valid

Add unit test for some error cases. The unit test coverage is back to 100%.

### Rationale

These a basic configuration option related backend host for the content cache.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: tests/unit/test_state.py [hunk 1]
```diff
@@ -208,11 +259,150 @@ def test_config_protocol_invalid():
...
+    charm.config[PROXY_CACHE_VALID_CONFIG_NAME] = '["200 302 30m", "400 1m", "500 1m"]'
+
+    config = Configuration.from_charm(charm)
+    assert config.hostname == "example.com"
+    assert config.path == "/"
+    assert config.backends == (IPv4Address("10.10.1.1"), IPv4Address("10.10.2.2"))
+    assert config.protocol == "https"
+    assert config.health_check_interval == 30
+    assert config.backends_path == "/"
+    assert config.proxy_cache_valid == ("200 302 30m", "400 1m", "500 1m")
```
Comment (jdkandersson) [inline tests/unit/test_state.py]: Do you think these could just be parametrized tests that expect valid and invalid values? ALso make sure we check edge cases thoroughly

**Comment by yhaliaw** [comment]
Still working on adding more integration tests.

File: src/state.py [hunk 2]
```diff
@@ -85,21 +93,63 @@ def validate_hostname(cls, value: str) -> str:
...
             The value after validation.
         """
-        # This are the valid characters for path in addition to `/`:
-        # a-z A-Z 0-9 . - _ ~ ! $ & ' ( ) * + , ; = : @
-        # https://datatracker.ietf.org/doc/html/rfc3986#section-3.3
-        valid_path = re.compile(r"[/\w.\-~!$&'()*+,;=:@]+", re.IGNORECASE)
-        if valid_path.fullmatch(value) is None:
-            raise ValueError("Path contains non-allowed character")
+        for item in value:
+            tokens = item.split(" ")
```
Comment (yanksyoon) [inline src/state.py:146]: Nit: Would whitespace separator not suffice here? (Would you be able to add the valid string example format just in case perhaps later as a follow up)

**Review by yanksyoon — DISMISSED** [review]
LGTM!

File: tests/unit/test_state.py (hunk 1 above)
Comment (yhaliaw) [reply tests/unit/test_state.py]: I have parametrized the tests and added more test cases with the valid values.

File: src/state.py (hunk 2 above)
Comment (yhaliaw) [reply src/state.py:146]: The format is defined by nginx. For single space as separator, I have tested and confirmed it works.
I am not sure if other whitespace or multiple whitespace would work with the nginx configuration.

I can add this a backlog item, but I think this is not high-priority right now.

File: config.yaml [hunk 3]
```diff
@@ -25,3 +25,23 @@ options:
...
+    type: string
+    default: /
+  proxy-cache-valid:
+    description: |
+      A list of HTTP response code(s) followed by a cache valid duration. For example ["200 302 1h
+      "", "404 1m"], which would mean 200 and 302 responses are cached for 1 hour and 404 responses
+      are cached for 1 minute. This configuration is passed to nginx proxy-cache-valid
+      (https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_valid).
+    type: string
+    default: '[]'
```
Comment (cbartz) [inline config.yaml:47]: I have not read the specification, but I assume this is covered or agreed.

File: src/state.py [hunk 4]
```diff
@@ -121,8 +171,24 @@ def from_charm(cls, charm: ops.CharmBase) -> "Configuration":
...
+        try:
+            proxy_cache_valid = json.loads(proxy_cache_valid_str)
+        except json.JSONDecodeError as err:
+            raise ConfigurationError(
+                f"Unable to parse proxy_cache_valid: {proxy_cache_valid_str}"
+            ) from err
+        if not isinstance(proxy_cache_valid, list):
+            raise ConfigurationError(
+                f"The proxy_cache_valid is not a list: {proxy_cache_valid_str}"
+            )
```
Comment (cbartz) [inline src/state.py]: Have you tried passing the `proxy_cache_valid_str` directly to the basemodel? Iirc, pydantic has some magic in converting / validating strings to specific objects.

File: src/state.py [hunk 5]
```diff
@@ -131,6 +197,9 @@ def from_charm(cls, charm: ops.CharmBase) -> "Configuration":
                 backends=backends,  # type: ignore
                 # Pydantic allows converting str to a string enum.
                 protocol=protocol,  # type: ignore
+                fail_timeout=fail_timeout,
+                backends_path=backends_path,
+                proxy_cache_valid=proxy_cache_valid,  # type: ignore
```
Comment (cbartz) [inline src/state.py:201]: Can you add explanatory comments for the ignore comments.

File: src/state.py [hunk 6]
```diff
@@ -170,3 +239,67 @@ def to_integration_data(self) -> dict[str, str]:
...
+def check_nginx_time_str(time_str: str) -> None:
+    """Check if nginx time str is valid.
+
+    Args:
+        time_str: The time str for nginx configuration.
+
+    Raises:
+        ValueError: The input is not valid time str for nginx.
+    """
+    time_char = set(("h", "m", "s"))
```
Comment (cbartz) [inline src/state.py]: optional: I prefer
```suggestion
    time_char = {"h", "m", "s"}
```

File: tests/unit/test_state.py [hunk 7]
```diff
@@ -190,6 +223,25 @@ def test_config_backends_invalid_backends(invalid_backends: str, error_message:
...
+    """
+    arrange: Mock charm with valid configurations.
+    act: Use a http as protocol, and create the configuration from the charm.
+    assert: Correct configurations from the mock charm.
+    """
+    charm = MockCharmFactory()
+    charm.config[PROTOCOL_CONFIG_NAME] = "http"
+
+    config = Configuration.from_charm(charm)
+    assert config.hostname == "example.com"
```
Comment (cbartz) [inline tests/unit/test_state.py:241]: nit: have a newline to separate the act from the assert block

File: src/state.py [hunk 8]
```diff
@@ -85,21 +93,63 @@ def validate_hostname(cls, value: str) -> str:
...
+            value: The value to validate.
+
+        Returns:
+            The value after validation.
+        """
+        return validate_path_value(value)
+
+    @pydantic.field_validator("proxy_cache_valid")
+    @classmethod
+    def validate_proxy_cache_valid(cls, value: str) -> str:
```
Comment (cbartz) [inline src/state.py]:
```suggestion
    def validate_proxy_cache_valid(cls, value: tuple[str, ...]) -> str:
```

**Review by cbartz — COMMENTED** [review]
Thanks. Have some (mostly minor) comments.

File: config.yaml (hunk 3 above)
Comment (yhaliaw) [reply config.yaml:47]: This is covered by ISD-178.
The `fail-timeout` is a new addition after discussion with David and Gergory today, but the name is not discussed in the discussion. I choose the name as that is what the nginx configuration is called.

File: src/state.py (hunk 8 above)
Comment (yhaliaw) [reply src/state.py]: Done.

File: src/state.py (hunk 6 above)
Comment (yhaliaw) [reply src/state.py]: Done.

File: tests/unit/test_state.py (hunk 7 above)
Comment (yhaliaw) [reply tests/unit/test_state.py:241]: Update the file to comply with this.

File: src/state.py (hunk 4 above)
Comment (yhaliaw) [reply src/state.py]: Tried it and it did not work.
It might be the proxy_cache_valid is of Tuple type.

**Review by jdkandersson — DISMISSED** [review]

**Review by cbartz — DISMISSED** [review]

File: src/state.py [hunk 9]
```diff
@@ -21,6 +21,9 @@
 PATH_CONFIG_NAME = "path"
 BACKENDS_CONFIG_NAME = "backends"
 PROTOCOL_CONFIG_NAME = "protocol"
+FAIL_TIMEOUT_CONFIG_NAME = "fail-timeout"
+BACKENDS_PATH_CONFIG_NAME = "backends-path"
+PROXY_CACHE_VALID_CONFIG_NAME = "proxy-cache-valid"
```
Comment (alithethird) [inline src/state.py:26]: Nitpick: These are named `*_FIELD_NAME` in the other charm. Would be nice to keep it standard in both charms.

**Review by alithethird — COMMENTED** [review]
Looks good. Just 1 nitpick.

**Comment by gregory-schiano** [comment]
@yhaliaw @jdkandersson This PR has been reviewed and previously approved by David, should I review it again ?

**Review by jdkandersson — APPROVED** [review]

File: src/state.py [hunk 10]
```diff
@@ -42,14 +45,19 @@ class Configuration(pydantic.BaseModel):
...
+        fail_timeout: The time to wait before using a backend after failure.
+        backends_path: The path to request the backends.
+        proxy_cache_valid: The cache valid duration.
     """
 
     hostname: typing.Annotated[str, pydantic.StringConstraints(min_length=1)]
     path: typing.Annotated[str, pydantic.StringConstraints(min_length=1)]
     backends: tuple[pydantic.IPvAnyAddress, ...]
     protocol: Protocol
+    fail_timeout: typing.Annotated[str, pydantic.StringConstraints(min_length=1)]
```
Comment (alithethird) [inline src/state.py:58]: Just an alternative, you can also call the validator in the `Annotated`
ex: 
```python
fail_timeout: typing.Annotated[str, pydantic.StringConstraints(min_length=1), BeforeValidator(check_nginx_time_str)]
```

https://docs.pydantic.dev/latest/concepts/validators/#annotated-validators

**Review by alithethird — APPROVED** [review]

**Comment by yhaliaw** [comment]
The publish-docs was not passing due to the main branch has wrong URL in the metadata.yaml.
David has confirmed force merging should fix this issue.
